plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs

```diff
@@ -165,6 +188,52 @@ lensResolveProvider _ ideState plId cl uri rd@(ResolveOne _ uid) = do
 lensResolveProvider _ _ _ _ _ rd = do
    throwError $ PluginInvalidParams (T.pack $ "Unexpected argument for lens resolve handler: " <> show rd)
 
+
+-- | Provide explicit imports in inlay hints.
+-- Applying textEdits can make the import explicit.
+-- There is currently no need to resolve inlay hints,
+-- as no tooltips or commands are provided in the label.
+inlayHintProvider :: Recorder (WithPriority Log) -> PluginMethodHandler IdeState 'Method_TextDocumentInlayHint
+inlayHintProvider _ state _ InlayHintParams {_textDocument = TextDocumentIdentifier {_uri}, _range = visibleRange} = do
+    isIHSupported <- liftIO $ isInlayHintsSupported state
+    if isIHSupported
+    then do
+        nfp <- getNormalizedFilePathE _uri
+        (ImportActionsResult {forLens, forResolve}, pm) <- runActionE "ImportActions" state $ useWithStaleE ImportActions nfp
+        let inlayHints = [ generateInlayHints newRange ie pm
+                         | (range, int) <- forLens
+                         , Just newRange <- [toCurrentRange pm range]
+                         , isSubrangeOf newRange visibleRange
+                         , Just ie <- [forResolve IM.!? int]]
+        pure $ InL inlayHints
+    -- When the client does not support inlay hints, fallback to the code lens,
+    -- so this is Null
+    else do pure $ InR Null
+  where
+    -- The appropriate and intended position for the hint hints to begin
+    -- is the end of the range for the code lens.
+    --   import Data.Char (isSpace)
+    --   |--- range ----|-- IH ---|
+    --                  |^-paddingLeft
+    --                  ^-_position
+    generateInlayHints :: Range -> ImportEdit -> PositionMapping -> InlayHint
+    generateInlayHints (Range _ end) ie pm =
+      InlayHint { _position = end
+                , _label = InL $ mkLabel ie
+                , _kind = Just InlayHintKind_Type -- for type annotations
+                , _textEdits = fmap singleton $ toTEdit pm ie
+                , _tooltip = Nothing
+                , _paddingLeft = Just True
+                , _paddingRight = Nothing
+                , _data_ = Nothing
+                }
+    mkLabel :: ImportEdit -> T.Text
+    mkLabel (ImportEdit{ieResType, ieText}) =
+      let title ExplicitImport = abbreviateImportTitle . T.dropWhile (/= '(') $ ieText
+          title RefineImport   = "Refine imports to " <> T.intercalate ", " (T.lines ieText)
```

Comment:
This seems odd. Do we have any tests showing this? These perhaps _shouldn't_ be shown as inlay hints? I'm unsure.

Reply:
It's possible these need to always be code lenses, since they don't represent implicit content that can be added to the document by accepting it, which is pretty much what inlay hints are supposed to be. Maybe I'm wrong though, not sure.

Reply:
Well, I was going to support refine with inlay hints, it may looks like:

`import qualified RefineA <inlay>from RefineB</inlay> as RA`

or:
`import RefineD <inlay>from RefineE</inlay> <inlay>( e2 )</inlay>`

But this is a little tricky for me, I'm trying to locate where the module name position is. In addition, I don't know if I understand it wrong, there may be _many_ original modules? This makes inlay hints longer and jagged.
So I want to ask your opinion.

Reply:
Yes, I think it would be pretty sensible to leave refine as a code lens and not an inlay hint. Let's do that for now and maybe open an issue to discuss what the best UX would be for refine imports?